### ISSUE TITLE
[FR][RFC] Build a serving framework to host and serve trained PyTorch models

### ISSUE THREAD

**Issue by vdantu** [pytorch/pytorch#27610]
## 🚀 Feature
Request to build a Serving framework to host and serve trained PyTorch models. 
@alexwong

## Motivation

PyTorch provides an excellent and easy-to-use interface to train a model and also provides an easy to use optimized inference interface through JIT. Currently, there is a need for an optimized model serving solution to take any model trained using PyTorch framework into production. There exists multiple solutions to serve a PyTorch model in production, but most of the solutions are generic model serving solutions. There are multiple pain points that the PyTorch community currently face when trying to take a PyTorch model into production. 

* Building a high performance web serving component to host PyTorch models is difficult to build and requires experience and domain knowledge.
* Adding custom pre-processing and post-processing for a model in service currently requires significant rework on the model server itself.
* Supporting multiple accelerators requires additional work.
* Any customization to the model server would require significant understanding of the existing serving framework itself and would also require significant rework.

We think the following are the most important requirements of a good PyTorch model serving framework.

## Pitch

We want to contribute to building a serving framework, which addresses the above pain points and more. We foresee the 

A PyTorch model serving solution will have the following capabilities: 

1. **Performance**: The server component should be highly-performant with low overhead from the serving framework. This implies that the average throughput must be high and the P90 latencies should be low. Its also important to get P50 and P90 latencies comparatively flat, signifying that all the requests are treated equally.
2. **Host Multiple Models**: The server component should be able to host multiple models at the same time and customers should be able to load/unload a model at runtime. The model serving framework should expose an endpoint for each model, which can be reached by any customer.
3. **High Availability**: The serving framework should be robust. Runtime errors of one model shouldn’t affect other models running on the server or the runtime of the server itself. There should be mechanisms to recuperate from any system out of resource errors.
4. **Metrics and Logs**:  A production grade serving framework should provide insight into the runtime of the model. The serving framework should provide easy access to logs and metrics and also provide easy hooks to add new logs and metrics without needing to understand the serving framework at a deep level.
5. **Support both Eager mode and Scripted mode models**: The serving framework should support means to run PyTorch models in scripted mode for optimized execution.
6. **Support for multiple bindings**: The serving framework should have support for models loaded via Python (eager/torchscript) or C++ bindings (JIT IR traces).
7. **Supports HTTP and gRPC Endpoints**: The serving framework should come with a full set of HTTP endpoints for managing models as well as running inference on the models. PyTorch serve would also come with an SDK to easily customize the endpoints. The serving framework would also support gRPC endpoints. 
8. **Ease of use and access**: The serving framework should be easy to set up on any platform (MacOS, Linux, Windows) and should be testable on these systems. Users should have the same experience to containerize the Serving framework and launch into production using any container orchestration mechanism. The PyTorch serve framework would also have a fully featured CLI to start and run the model server.
9. **Lightweight**: This implies that the serving component itself shouldn’t have multiple dependencies.
10. **Supports features such as request batching**: The serving framework would have features such as request batching, to optimally run inference on accelerators.
11. **Support model Versioning and A/B testing**: The serving framework should have capabilities to load multiple versions of the same model and run A/B tests on the model. This is very useful for when rolling out newer versions of a model into production. This can also be used to roll back if the new model is not as performant. 
12. **Zero code serving**: While providing the feature to customize the pre-processing and post-processing of inference requests, the PyTorch serve framework should also allow customers to simply drop their trained models into the server and use it for inference. In other words, the PyTorch serving framework should come with sensible defaults for pre processing and post processing. 
13. **Easy customizability**: The serving framework must be easy to customize for endpoints. This means easily modifying and adding new management endpoints, defining custom request batching algorithms, defining custom AuthZ and AuthN mechanisms. 
14. **Support Accelerators**: A production grade model server should be able to run on GPU hosts as well as any other custom accelerator host. 
15. **Web UI**: The PyTorch serving framework should come with a Web UI to allow interaction with a served model.

### Proposed Technical Details

The model server will be based on a micro-service based solution rather than a monolithic approach. This architecture would bring us the benefit of decoupling the work of handling ingress requests and running the actual inference. This also allows the PyTorch serving framework to scale beyond a single host. The high-level components of the serving framework are divided into a frontend and backend which share different responsibilities.

#### Frontend responsibilities:

1. **Manage connections**:  In other words, the incoming requests and outgoing responses are managed by the frontend.
2. **Manage models** : The frontend is responsible for the lifecycle of the model. Each hosted model will have its unique endpoint created, which could take any data type and return any data-type back. 
3. **Manage the backend**: The frontend is responsible for providing the models to be loaded onto the backend workers and also managing the backend workers themselves. 
4. **Manage requests**: Requests coming into the server’s frontend will be queued in model-sepecific queues for handling. 
5. **Request distribution**: The frontend will be responsible for request distribution, to backend workers. 
6. **Metrics and Logs** : Frontend will be responsible for metrics management, logs management and capturing any custom metrics and logs that come from the backend.  
7. **Retrieve models from anywhere** : The frontend is also responsible for retrieving the models from cloud or local storage. 

#### Backend responsibilities:

1. **Running inference**: Tightly integrate with PyTorch backend and also responsible for running any preprocessing required, running the *forward* method of the model with the incoming request and running post process on the inference response. 
2. **Default pre-process, inference and post-process**: If no custom processing logic is provided to the backend, it would have default logic to run preprocess, inference and post process on the model.
3. **Publish custom metrics and logs**: Backend will have the capabilities to publish custom model level metrics and logs.

#### Monitoring status of the Server
![monitoring](https://user-images.githubusercontent.com/36211508/66499650-153e7a00-ea75-11e9-9d41-31a00b89452f.png)

#### Checking status of the models
![Model status](https://user-images.githubusercontent.com/36211508/66499662-196a9780-ea75-11e9-9275-83337104c3de.png)

#### Running inference
![Inference](https://user-images.githubusercontent.com/36211508/66499668-1cfe1e80-ea75-11e9-98ef-00120b268104.png)

#### Loading model
![](https://user-images.githubusercontent.com/36211508/66499722-33a47580-ea75-11e9-92e3-0658e57b1bae.png)

#### Deleting a model
![delete_model](https://user-images.githubusercontent.com/36211508/66499725-356e3900-ea75-11e9-8e5a-ae032e231e69.png)=

## Next Steps

* We are looking for feedback/comments on this proposal. Specifically, we are looking for feedback on the list of capabilities outlined in this RFC and their priority. We also welcome feedback on our proposed design and implementation. 
* Add details on proposed architecture and details on endpoint.
* Add additional sequence diagrams.
* Target Q4 2019 for Experimental release.

**Comment by praateekmahajan** [comment]
Would it be fair to say, that on the front end, there should some sort of checkbox that asks the user if they want to : 
1. Quantize the models
2. Do layer fusion

These functions would be really cool, if supported inherently by torchserve.

**Comment by soumith** [comment]
I'd like to see some more details on the following aspects:

1. Model lifecycle / serving multiple models at the same time:
  - What is the mechanism with which you shall do this?
     - Will you simply dereference the pointer to the old model and load in the new model?
     - Will you push the old model to a cheaper memory space such as CPU memory, in case it has to be loaded back into GPU memory again?
      Please flesh out details
  - Serving multiple models at the same time:
     - how is memory managed and partitioned?
     - How will latency guarantees be given in such a case?

2. Flesh out more details about how you would do "zero-code" preprocessing / post-processing. Give example use-cases and what happens in each of them. Give API details where necessary.

**Comment by suo** [comment]
I'm curious how you envision how this interacts with model management overall. One problem that is a continual source of pain for us internally is that users are encouraged by the design of our internal serving framework to work with their models as data blobs (a `.pt` file in this case) instead of code.

This complicates maintenance, as you can get into a state where there are models running on the server that cannot be rebuilt and redeployed, because the original model code has long since disappeared. This makes it difficult to do framework upgrades, as we can't "codemod" old models to make the upgrade safe in cases of BC breakage. It is akin to shipping off a binary and not keeping around the source code so that it can be changed, receive security and feature upgrades, and so on.

I'm concerned that if the API is designed around "give us a model `.pt` file and we will commit to serving it potentially forever", we will be indirectly encouraging these bad practices. One solution is to have the API be: give us a Python program that, when run, produces a `.pt` file and hydrates it with the appropriate weights.

**Comment by vdantu** [comment]
@praateekmahajan @soumith @suo : Sorry for delay in responding. Firstly, thanks for your comments and interest. We will work on detailing the internals and update the RFC. We will also detail the model management in detail. 

As mentioned above, we will provide more details of the internal design, but just to touch upon some of the concerns,
1. We do envision supporting eager mode python script, ```.pt``` file and JIT mode for execution. We will try to capture how this can be done. 
2. @suo: I wanted to understand more on the bad practice that you were alluding to. Could you explain why running a model for longer in production service is a bad practice? 
3. @soumith : You touched upon a very good point. The model server is envisioned to handle out of memory gracefully. We also envision that for the first stage of the project that the orchestration layer is outside of the model server. The server would provide necessary primitives to ease scaling/deleting orchestration.  
4. @soumith : I think your question on where the old model goes is a part of model-versioning question. I would assume that old models should be decommissioned once the new model version is deemed good. Is this assumption wrong? 

Again, thanks for asking these questions. We will try and get the details out soon.

**Comment by nairbv** [comment]
A couple of thoughts:

Related to @suo's comment, I think we want to be able to distinguish the model from model parameters (`state_dict`) in model management, as described here:
https://pytorch.org/tutorials/beginner/saving_loading_models.html
Doing so has implications for the design of the model management API and how we handle versioning, since we'd want to be update a model's code or state_dict independently.

I'd like to see more details on the proposed SDK. I'm naively imaging that in python this might look something like:
```
torchserve.publish(myPytorchModel, "modelName", service_url)
...
model_reference = torchserve.model("modelName")
resultTensor = model_reference(inputTensor)
```
but there are a lot of details that could be flushed out. In this example I'm assuming no server-side pre/post processing other than tensor serialization. What would the versioning APIs look like? Maybe we can fetch "latest," "stable," or specific version IDs of a model?

Do we expect standard/configurable pre-processors to be built into the model server, e.g., for images, or would they always be customized per-model? I could imagine a user wanting to be able to pass either an image or already-processed tensor to the same model without deploying two instances of that model.

Since this is gRPC, would the tensor serialization format be protobuf and/or json for raw inference tensors? Looking at the inference sequence diagram, the "{yes}" from "backend pytorch" implies post-processing in python. In cases where there is no post-processing (if the result is a raw tensor), considering both python and C++ inference, does the backend need to be able to serialize tensor results to return to the model server?

**Comment by jspisak** [comment]
cc @vdantu - can you address these questions? cc @vinarm

**Comment by vdantu** [comment]
@nairbv : Hi Brian,
Thanks for sharing the link about state_dict. 
As I understand, its very difficult to define a _default load-model functionality_ with eager mode. I think this is understandable. We foresee that the models , ie., state-dict and/or the eager mode script, would be brought into the TorchServe at model loading time. 
I am still not clear about one of the points you make,
***Doing so has implications for the design of the model management API and how we handle versioning, since we'd want to be able to update a model's code or state_dict independently.***
Both code and the state-dict can be brought independently into TorchServe, in the above method. But, the model code on the running TorchServe will be updated only after reloading/restarting the running instance of the model. I would assume that a new version endpoint should be created only when there is a code-change. Is this assumption wrong? 

The SDK I had mentioned was not the SDK to launch the TorchServe. Do you see a value in having an SDK to launch TorchServe? Considering that TorchServe would be a server process, we propose that it should have a CLI. So, to start the server process customer UX would be, 
```
$ torchserve {OPTIONS} {MODELS}
```
The client SDK would just send REST/gRPC based requests to the TorchServe server on a particular ***IP:PORT/ENDPOINT***. 

The SDK I am talking about is the mechanism in which customers can provide custom pre-processing, load-model, inference and post-processing logic. This SDK will also have some common **defaults** to provide certain pre-processing, load-model, inference and post-processing  functionality. **_I think the SDK should be a separate FR as it requires more focussed discussions_**.

***Do we expect standard/configurable pre-processors to be built into the model server, e.g., for images, or would they always be customized per-model? I could imagine a user wanting to be able to pass either an image or already-processed tensor to the same model without deploying two instances of that model.***
Yes, as a standard server, it should be able to handle multiple data types. The SDK I mentioned above should be able to provide some standard defaults. Customers would be able to define their own custom pre-processing and post-processing as well. We **shouldn't** require a new model-deployment to handle different incoming request types. The same launched instance of the model should be able to service difference request types, such as tensors or images (in case of image based models).   

Regarding gRPC support. We propose we start with a REST based interface and then extend it to gRPC interface. As you mentioned the serialization could be JSON based. Would customers find it beneficial to be able to send pickle tensors to remote server for inference? 
You are correct that the backend serializes the response. The response in the above example is similar to "application/text" but, customers should be able to send back binary blobs of data or JSON data or any other format data.

**Comment by nairbv** [comment]
> We foresee that the models , ie., state-dict and/or the eager mode script, would be brought into the TorchServe at model loading time.

In this context I think by "loading" you mean sending/publishing the model to the server, is that correct? We might want to refine our naming since `torch.load` is a PyTorch function that would load a model into python from disk. We may have other ambiguous terms to define.

> I would assume that a new version endpoint should be created only when there is a code-change. Is this assumption wrong?

This may be another example where we could better define terms. I think I agree that if someone wants to update the state-dict of a model, they would typically want to do so without creating a new "model endpoint," but I think they would be creating a new "version" of that endpoint. Of course nothing stops someone from publishing one model twice to create two different endpoints.

Concretely, I'm imagining a REST API where a user might create a model by posting to something like `POST /api/model/create?name=MyModel`, and get a response like `{model: { name: "MyModel", model_id: 1234, model_version: 1 } }`. Here by "model endpoint" I'd be referring to something like `/api/model/1234`, and by model version I mean the `1`. If the user posts an update to a model endpoint that already exists (`/api/model/1234`), maybe they'd get a response with `model_version: 2`. To run inference, maybe a user would then post an image with something like `POST /api/model/1234/inference?model_version=2&preprocess_type=image`, but could also hopefully run inference using the latest version of a model without having to specify a version ID.

This is just my guess at how I imagine something like this might work, but I'd like to see more detail of how you're thinking about versioning and how exactly the API would work.

It might be useful to have the model server be capable of fetching models and state dicts specified from a URL or model library instead of always needing to post the models to the server.

Related to #11 above, 
> 11. Support model Versioning and A/B testing

It might help to flush out the details of what it might look like by thinking through a use-case, since there are probably other ways to handle it, and I may misunderstand the requirements we're trying to handle.

I'm imagining a workflow where I've published a trained model, I re-train it with newer data, and want to publish the updated model. If the code hasn't changed, I'll only want to publish the new state-dict, though I could also imagine tweaking some model code. Semantically it's still the same model since it serves the same inference purpose. I probably don't want to change production code that relies on this model, but I also still might also want to be able to run the original version too for A/B testing.

I could imagine being able to refer to either a specific model version, or the "latest" or "stable" versions of a model. I'd hope production code referring to the "stable" version of the model could seamlessly continue getting good inference values, and wouldn't need to change any code when I called some API endpoint that bumped 'stable' to point to an updated version.

> But, the model code on the running TorchServe will be updated only after reloading/restarting the running instance of the model.

It would probably be desirable to be able to publish an updated version of a model to the same endpoint with different code, and to be able to run both versions of the model simultaneously. It sounds like you're saying this wouldn't be a supported use-case? Is this a technical limitation?

> The SDK I had mentioned was not the SDK to launch the TorchServe

I didn't mean an SDK for launching a server either, I meant a client SDK for model publication and inference. E.g. I could picture (from python code) defining a PyTorch model and publishing it to the server, then running inference using a reference to that published model. This could be simpler for developers vs needing to manually manage models on disk and directly call the REST API to update/use the served model, and then needing to parse some JSON response back into a tensor object.

I'm thinking a python client SDK that might have operations like: list published models, list versions of a model, get a a reference to a remote model, run inference on a remote model with a python input and tensor output, publish a new model, update the state-dict of a model, update tags on some version of a model (e.g. updating "stable" to point a specific version), etc.

Torch Hub isn't a model server, but in a similar sense Torch Hub has a library for publishing and downloading models:
https://pytorch.org/docs/stable/hub.html

> You are correct that the backend serializes the response.

How would this part work? If "backend PyTorch" in your diagram just refers to PyTorch itself, I don't think PyTorch has built-in mechanisms to return a tensor as json or text, and presumably there'd need to be some consistency between Python/C++ model inference. Does this requires changes to PyTorch?

**Comment by vdantu** [comment]
@nairbv 

> In this context I think by "loading" you mean sending/publishing the model to the server, is that correct?

Yes. You are correct. I agree, the terminology is confusing. I meant registering the model with server.

>This is just my guess at how I imagine something like this might work, but I'd like to see more detail of how you're thinking about versioning and how exactly the API would work.

Yes. I was also thinking in the same lines for the REST endpoints. We will take an **action item** and publish this doc with the REST APIs soon. We could discuss over that document. 

> Support model Versioning and A/B testing
> It might help to flush out the details of what it might look like by thinking through a use-case, since there are probably other ways to handle it, and I may misunderstand the requirements we're trying to handle.

This definitely needs further discussion. I would recommend raising a separate FR for this and working with it. I think in my document about APIs I will also list the features which I believe should be separate FRs and separate discussions.  

> It sounds like you're saying this wouldn't be a supported use-case? Is this a technical limitation?

I am not clear on what it means to load the changed code to the same model. Technically you would be able to. Either the code running needs to read updated state-dict, this can be done at runtime. But, if  the code changed, then we would need to reload the model because the running instance of the model has changed. We should support both. It would be a manual process in the first iteration. Customers will have to manually restart the endpoint (to update the code). I think this manual step also keeps things stable in production and avoids automatic unintended code updates from happening. 

> This could be simpler for developers vs needing to manually manage models on disk and directly call the REST API to update/use the served model, and then needing to parse some JSON response back into a tensor object.

I completely agree with you. This will be a very valuable add. I see this as a glue layer between users and the actual server. Let me note this down as well in the requirements. We should have a separate FR for this. 

I quickly drew a depiction of how this would look like. 

![image](https://user-images.githubusercontent.com/36211508/68441065-ca05ad00-0181-11ea-8a15-187e2a986f78.png)

The highlighted SDK is responsible for providing some defaults to converts inputs and outputs based on the content-type/accept-type headers. This SDK will also be able to take any custom pre-processing or post-processing that customers would want to provide, if customers don't want to use the defaults. This SDK is something we should discuss more on. **Action item**: We will send out a document detailing the design and we can have a discussion. 

In the long run, if we feel that this SDK is unnecessary and that it can be upstreamed into PyTorch, repo we could do that as well as a separate effort. 

As next steps, I think we will work on the above "**action items**"

**Comment by soumith** [comment]
I still don't understand how an eager mode model is loaded into TorchServe. The weights are clearly loaded with `model.load_state_dict(torch.load('weight.pt'))` and that's well and good. But how exactly is `model` loaded into TorchServe, and how exactly is it versioned?
I'd like to see a concrete API and usage, because it's not obvious at all on how this works.
As a simple example, if `model` here refers to some `MyComplexTransformer` python class, which is defined in a `.py` file, and in itself has a bunch of package dependencies, how exactly is this whole thing versioned and reliably loaded?

The solution in the case of JIT / TorchScript models is obvious, as the graph and weights are stored as a single package with no dependencies except pytorch.

Also, if you could expand on the details of the memory management model, it'd be helpful.

> We also envision that for the first stage of the project that the orchestration layer is outside of the model server. The server would provide necessary primitives to ease scaling/deleting orchestration.

What exactly does this mean, with a code illustration.

Looking forward to the detailed document with design. Please feel free to include detailed API / usage in the design document.

**Comment by vdantu** [comment]
@soumith : I will include the UX, pseudo-code illustration and pointers on memory management in the document as well. I will try and publish the document first thing next week.

**Comment by nairbv** [comment]
> if the code changed, then we would need to reload the model because the running instance of the model has changed. [...] Customers will have to manually restart the endpoint (to update the code).

I'm not sure I understand this. My expectation would be that the running model continues to run, but that the new version of the model with new code can also run. Which one gets used by a particular inference call would depend on which version the user requested.

**Comment by vdantu** [comment]
@nairbv : Yes. You are correct. We can run multiple versions of the same model in parallel, on different endpoints. But, if a customer wants to deploy modified code to an already running/existing endpoint, that endpoint needs to be redeployed.

**Comment by nairbv** [comment]
> @nairbv : Yes. You are correct. We can run multiple versions of the same model in parallel, on different endpoints. But, if a customer wants to deploy modified code to an already running/existing endpoint, that endpoint needs to be redeployed.

An "endpoint" can be parameterized though right? Are you considering, say, `/model/1234?version=1` to be a different endpoint from `/model/1234?version=2`? I usually think of an endpoint as being the path without the parameters. I would imagine that differing parameters at the same endpoint could point to two different running versions of a model.

**Comment by vdantu** [comment]
Design wise, it might be cleaner to have versioning as a part of endpoint. Though technically, both can be supported. I think asking customers to give a key-value as query parameters increases barrier for adoption. I think if model-id (which is a variable) is a part of the endpoint name, then version number (which is related to a particular model/model-id) should also be a part of the endpoint name. Also, query-parameters look optional. It makes enforcing version number hard. Do you see any advantage of having a separate query-parameter to define versions?

**Comment by nairbv** [comment]
> Design wise, it might be cleaner to have versioning as a part of endpoint. 

There are many ways it could be implemented, I don't have a strong preference, it's just not clear to me how specifically the use case is handled.

Regardless of if it's in the endpoint or a query parameter, we should be able to refer to a version of a model before and after a state dict or code change, and I'd think neither should require a reload.

> It makes enforcing version number hard.

The use-case I'm asking about is when a user wants to run a test client that always uses the "latest" version of a model, or a prod client that always uses the most recent tested/approved/stable version of a model. 

If the user is required to explicitly specify one version ID for a model in their client, then absent some workaround, the client code or configuration would need to be redeployed whenever the model parameters were updated. I'm not arguing that it should be /mymodel/?version=5 instead of /mymodel/version/5, just that it would be nice to be able to say version=latest where "latest" refers to a different version over time. I could imagine there being a default, which is why I was thinking of it as  an parameter, but there are other ways it could be done.

**Comment by fbbradheintz** [comment]
> Related to #11 above,
> 
> > 1. Support model Versioning and A/B testing
> 
> It might help to flush out the details of what it might look like by thinking through a use-case, since there are probably other ways to handle it, and I may misunderstand the requirements we're trying to handle.
> 
> I'm imagining a workflow where I've published a trained model, I re-train it with newer data, and want to publish the updated model. If the code hasn't changed, I'll only want to publish the new state-dict, though I could also imagine tweaking some model code. Semantically it's still the same model since it serves the same inference purpose. I probably don't want to change production code that relies on this model, but I also still might also want to be able to run the original version too for A/B testing.

Chiming in here, because I believe item #11 was based at least in part on my input, and I should probably make clear the sort of thing I was talking about.

Simple version support and A/B testing are two ends of a spectrum of possible behaviors to enable. There are certainly many steps along that spectrum.

**The most straightforward first step** is just loading new weights for an existing computation graph. This is the `state_dict` case discussed above, where only the learning parameters have changed, and is probably the simplest thing to implement. As I imagine the API `POST` call to `/model/1234` that verifies the number and arrangement of learning weights in the posted state dictionary and returned a new version number would be needed, and one should be able to access a particular version with a `GET` to `/model/1234?v=5`. We should think about the behavior of `GET /model/1234` in this case - that is, whether we want it to serve the latest, or whether we want a "preferred" version to be configurable, allowing regular inference to be done on model v4 in production while one tests v5 via the query string in the prod server env.

The issue I see with this is that it depends on having the model code that understands the state dict - probably the Python code used to develop the model. Great for eager mode, not so much for many prod use cases (because Python).

**The next level up** is to accommodate versions that change *both* the computation graph and the weights, without changing the semantics of the model. In eager mode, this means putting up code + weights. In graph mode, this just means putting up a new TorchScript file. Similar APIs to the previous case could suffice.

The sticky part here is that now you have two ways to define both "model" and "version" - Python code + state dict, or a single TorchScript export. *This suggests to me that we should think hard about the degree to which we wish to support eager mode in TorchServe.* There may be wins to be had with debuggability &c by supporting eager mode, but having these two different and complex instantiations (code + weights vs. TorchScript) of semantically similar objects (model versions taking identical inputs & outputs) sounds like a significant lift. I'd suggest that one way to mitigate that is to steer people onto our preferred "happy path" (TorchScript + JIT), while limiting support for eager mode.

**Another level up,** tangential to the idea of making a "preferred" served version configurable, is to allow the user to split traffic among versions - say, 80% to v4 and 20% to v5. You might want to do this as part of deploy new model versions, starting by testing the version manually, then serving a split as a response to the default URL for the purpose of observing behavior at higher scale. This leaves monitoring/measurement to the user, but gives them a tool to easily configure this behavior as part of a scaled deployment process. Certainly, the same could be accomplished by other means, such a proxy between the web request and the model server, but that's an extra layer of IT hassle.

**Yet another level up** would be A/B or multivariate testing: Allow the user to define such a test between 2+ model versions, have a way to pipe in measurements on accuracy, performance, or other metrics, monitor the differential results and their statistical significance, and have a way (manual or automated) to switch over to the better-performing model. This (or variants of it) are common practices at companies doing scaled web & web service deployments; a production inference service is still a service, and will have the same concerns about versioning and the understanding of differential performance in real-world use. Alternatively, instead of including this as a feature of the tool, we could add hooks that connect to tools for A/B testing services people are already using.

As to how all this relates to the first cut, IMHO:

* Versioning that includes changes to weights + graph of semantically identical models should be a *sine qua non* of an initial release, along with a way to give some control over which version gets served (i.e., not assuming it must/should be the latest). I favor the querystring approach.
* We should consider focusing on TorchScript deployments over eager mode, since it's where we want the community to go for production deployment.
* More complex behaviors (traffic splits, testing & cutover) should be included in the roadmap for future revs, with their implementation details and relative prioritization guided by community feedback and usage of the tool.
* Inclusion of testing features vs. hooks to other tools should be a decision made based on community feedback.

I'm very interested to hear how well everyone thinks this lines up with actual usage of serving solutions in the wild, and where we'd like it to go.

**Comment by vdantu** [comment]
We are still working on the document and are reviewing it. We will publish the document by end of the week.

**Comment by jusonn** [comment]
How will model server loads pytorch model?
Will JIT be essential to model server? or there will be another model object like tf saved_model??

**Comment by vdantu** [comment]
@jusonn : TorchServe will not rely only on JIT. We are proposing to support both JIT and Eager Mode scripts. 

In the initial release we want to provide a TorchServe serialization tool for customers who want to use eager-mode scripts. This tool can be used to serialize the eager mode model-script along with state_dict and/or any other extra files necessary to serve the model. This serialized model can be readily hosted on TorchServe and is also easily portable.

We understand that this might add one additional manual step during deployment, in the initial release, but we can work automating this step or including this step with the **TorchServe SDK** and eliminate this step. 

We are currently not looking into modifying the existing PyTorch serialization format.

**Comment by nairbv** [comment]
Another question I have is what this looks like when we scale to multiple hosts. I'm imagining multiple model servers running inference behind a load balancer, but then it's not clear how the models are managed across all inference hosts.

**Comment by vdantu** [comment]
I have published an architecture diagram above in the main description along with the API spec. Please share your views on this.

@nairbv : If the server is behind managed container orchestration service (and a load balancer). In my experience, in managed orchestrations, its easier to have individual cluster of containers host a single model. Then customers can have multiple of these model-server clusters hosting different models. Each of these clusters have their unique model inference paths. In this case, the model management would boil down to creation and removal of those container clusters. The loadbalancer could be configured with path-based-routing for inference, where different endpoint paths will point to different clusters/targets.

I have also seen services writing their own orchestration layer where the model-servers sit behind this orchestration layer. This orchestration layer understands the service SLAs. This layer is hence responsible for authz/authn, rate limiting the requests, allocation of system resources etc. In this case, the orchestration layer is responsible for managing the model-server itself. These services could use the dynamic loading/unloading of models that the model-server provides natively, to manage the models.

I am not sure I am simplifying things too much :). Please do chime in what you have seen in the past or what you think would simplify things. I would be very interested to learn more about your thoughts.

**Comment by nairbv** [comment]
> its easier to have individual cluster of containers host a single model.

This approach could require quite a bit more management though, in that it requires users to allocate at least one host to each model, and presumably either run a load balancer per model or add specialized configuration per path. I'm also not sure how it would address the issue of how to load a model on a new host before adding it to the cluster for that model.

> Each of these clusters have their unique model inference paths.

Each unique path would also at least need to be able to handle multiple versions of a model, since we're anticipating client code in production that would be able to continue to use a single "model" while it's implementation changes (i.e. when a new version/weights are deployed), as discussed above.

>  In this case, the model management would boil down to creation and removal of those container clusters.

I'm still not clear on how those models would get loaded onto new nodes when added to a cluster. Running one model per cluster doesn't necessarily solve that problem.

> I have also seen services writing their own orchestration layer...
> This layer is hence responsible for authz/authn, rate limiting the requests, allocation of system resources etc....
> These services could use the dynamic loading/unloading of models that the model-server provides natively, to manage the models.

I'm not sure what you mean here. I think we would want to ensure that we have a mechanism for scale-out to multiple hosts that doesn't require users to write a customized orchestration layer. I'd anticipate users wanting to run this behind a standard existing load balancer. I'm not sure how a standard load balancing system would manage the models without custom code, or where it would look up the models it would need to load on newly added servers.

> I am not sure I am simplifying things too much :). Please do chime in what you have seen in the past or what you think would simplify things. I would be very interested to learn more about your thoughts.

One possibly alternate solution/design:

The model server could be composed of two web services: an "inference" engine and a model management server (not sure what it would be called: could be "management server", "model hosting", "model library").

In a single-host deployment, the two services would be started on the same server (though maybe on different ports). In a larger deployment, the inference engines would be run behind a load balancer, and lazily-load models from the single-host "model management" server.

Some advantages:

* The user publishes models to a single host, but they can be queried on any inference host.
* The management server could use some simple IP or port based auth mechanism to control of model publication, while inference engines could be made more broadly available.
* Assuming some expiration mechanism, unused models or older versions of models that aren't frequently used wouldn't consume resources on any of the inference engines.
* Off-the-shelf load balancers usually support consistent hashing, which could keep models evenly distributed across inference engines without manual configuration.
* Should be resilient to failures or replacing instances in the cluster, since they'll just re-fetch any needed models from the management server. In a shrinking cluster, some instances get repurposed by fetching additional models as they handle more requests.

Some drawbacks of this approach would be:

* The first inference call for a model on a newly deployed inference engine could be slow while it lazily-fetches the model from the "management" server.
* Doesn't handle scaling the management server, though this may not be necessary.
* More complicated to manage two separate services.


Besides the scaled-out system design though, we also still have a few outstanding questions here, e.g. @soumith 's question on how model dependencies get loaded for non-jit models, and how pre/post processing and serialization work.

**Comment by vdantu** [comment]
@nairbv : Firstly, thanks for sharing your thought process in detail. I am really liking this discussion and seeing your perspective here.   
 
>I'm also not sure how it would address the issue of how to load a model on a new host before adding it to the cluster for that model.

Since we have only one model per cluster of containers running behind a loadbalancer, the addition of a new model would mean bringing up a new cluster of containers . The loadbalancer can be shared (depending on the QPS that the service wants to support) and we could add path based routing per target group/container cluster. 

We can possibly have a shared volume across these container clusters and let model-server run audit on this shared volume to update their models. But the audit mechanism would make the production setup flaky as these audit's can run at different times in different containers making the cluster inconsistent. I am not sure how we could broadcast model-management messages to all the containers in a cluster using a loadbalancer and synchronously get success from each of these containers. Do feel free to correct me if I am missing something here. 

>Each unique path would also at least need to be able to handle multiple versions of a model, since we're anticipating client code in production that would be able to continue to use a single "model" while it's implementation changes (i.e. when a new version/weights are deployed), as discussed above.

I understand what you are suggesting here. I think versioning is a valid requirement to be supported. But, customers who use loadbalancer and manage their model-servers using one of the managed container orchestration services (like Amazon ECS for example),  would need to create new endpoints for new models (or new versions of a model). Do you see how this can be done any differently? In this case, model versioning would be similar to having different endpoints for each version. This is in context of using model server in managed services. For customers writing their own orchestration logic, versioning in the API or the query string parameter will be more valuable.

> I'm not sure how a standard load balancing system would manage the models without custom code, or where it would look up the models it would need to load on newly added servers.

Yes, I am also foreseeing customers needing to write custom code to manage models in the first release of model-server. But, once we implement a client SDK for torchserve, this logic can be pushed into that client SDK. This Client SDK would be used to create model server cluster and also register models with the model server. This is shown in the architecture document as 
```
torchserve_client.deploy(...)
```

>The model server could be composed of two web services: an "inference" engine and a model management server

It seems like the client sdk in a web UI format? Is my thought wrong here? I didn't quite get your point **In a larger deployment, the inference engines would be run behind a load balancer, and lazily-load models from the single-host "model management" server**. The current architecture technically can scale beyond a single host, as all the communication is based on sockets. But, depending on the scope, we might be building container orchestration services inside the model server. We should probably have a separate micro service to make this happen.

>There could also be some way to build an architecture like this with eager loading to inference engines, but I'm not sure what that would look like.

Yes. We thought about this as well, in our initial design. We had the same concerns that you pointed out. Worse case scenario, the server might be in thrashing mode where old models are deleted and new ones are loaded. 

> ...question on how model dependencies get loaded for non-jit models, and how pre/post processing and serialization work.

In the initial release the customers would have to build the containers with the dependencies before launching it for inference. But, in the subsequent releases we can include the dependency management as a part of model-server and/or the client SDK. This gets into what the dependencies are (if they are native binaries that are required or the python pacakges that are required). Here, customers might need to provide some sort of `requirements.txt` or some other configuration that we define which give a list of packages that need be installed on the model server container before launching the model. Model server has a different PYTHONPATH for each of the python based models (non-jit) loaded on the model server. So we should be able to manage pip package dependencies here.

I don't foresee the current architecture needing significant change for any of these requirements.

**Comment by nairbv** [comment]
> Since we have only one model per cluster 

I think there are deeper issues I'm trying to describe that I don't think I'm getting across clearly, but this already seems fundamentally very hard on users. I could imagine a deployment where 5 servers would be enough compute to run inference on 10 models of varying utilization, and where each model might have a long history of versions. A user won't want to leave one server under-utilized, and may also want sufficient capacity for a model that requires more than one node.

The initial proposal includes "Host multiple models," "High availability" and "scale beyond a single host." I wouldn't expect the first two requirements to be mutually exclusive with the last.

Some models might also be regularly retrained as new data arrives, meaning there could be a new "version" of the model deployed and A/B tested quite frequently. I suspect many users would prefer to never delete old models/versions.

> the addition of a new model would mean bringing up a new cluster of containers
and
> But, once we implement a client SDK for torchserve, this logic can be pushed into that client SDK.

I don't think we can expect something like `torchserve_client.deploy(myModel)` to bring up a new cluster of containers or reconfigure a load balancer. Even if it did, that wouldn't solve the problem of how we add a new node to the cluster after having deployed a new model.

The proposal includes "Users should have the same experience to containerize the Serving framework and launch into production using any container orchestration mechanism." If we were saying they'd need to write their own orchestration system to manage models across nodes in prod, we'd be offering much less, and building such a custom orchestration system could be quite challenging. I don't think it would be possible to put that inside the sdk, since the sdk is client-side and unaware of server-side hardware failures etc.

> Do you see how this can be done any differently? In this case, model versioning would be similar to having different endpoints for each version. 

I think what you're considering here is that if /model/1234?version=latest first points to version=1 and later points to version=2, the load balancer could be directing requests for "latest", "1", and "2" to three different hosts or clusters of hosts.  The version could just be ignored by the load balancer. If the LB ignores the version when hashing, the hash would still send all queries for a particular model to the same (set of) servers. The total query load for a model is probably more stable than the query load for a specific version of a model anyway, so this is probably the desired LB configuration. When A/B testing, some percent of the total number of users might be directed to version A of a model, with the remainder on version B.

> I am not sure how we could broadcast model-management messages to all the containers in a cluster using a loadbalancer and synchronously get success from each of these containers.

Exactly. That's part of why I'm thinking this approach wouldn't satisfy the requirements. This would be difficult to build and I don't see a way to build it that wouldn't be prone to race conditions and stale unreachable host lists.

> It seems like the client sdk in a web UI format? Is my thought wrong here?

No, I meant a single repository of models that would be published-to and queried by the sdk, from which the inference engines could find model definitions and metadata. A server that is not behind a load balancer (or to which the load balancer directs all model management requests), which any server can directly query to get a full list of models and metadata available for serving.

If every inference node in the cluster can refer to one "management" host that has all of the model information, then we don't have to worry about orchestration/audit/etc. If any inference node can lazily fetch data about any arbitrary model/version from one central node, then it also doesn't matter what requests the load balancer directs to that host, or how many models are served by one host, simplifying load balancer configuration. Any inference host could handle any model just by checking the management server for the details.

> customers would have to build the containers with the dependencies

I think we're looking for the details of the format in which a non-jit model could be published so that it contains or refers to all its dependencies.

There are also the questions about serialization between the model server and pytorch, and how pre/post-processing fits in:

In your example you just say "data for inference" and a response of "yes," but in the "Zero code serving" case it's not clear how a pytorch tensor object or other input is going to be communicated to a generic "pytorch backend" without any special (de)serialization code. Is there some other layer that deserializes the tensor object before passing it to the pytorch model? If so I think that's not clearly detailed.

**Comment by alexwong** [comment]
Sorry just joining the convo now so forgive me if I am misinterpreting things!

> I think we're looking for the details of the format in which a non-jit model could be published so that it contains or refers to all its dependencies.

In a first release, we envision the user packaging their needed files (so both the model definition and parameters) when providing it to the model server. This could also include custom handler code for preprocessing and postprocessing and also dependent files(ie other classes and stuff). I think library dependencies (things you'd install w/ pip or build locally) would live outside of the model server for now and would require the user to directly handle on their host.

> In your example you just say "data for inference" and a response of "yes," but in the "Zero code serving" case it's not clear how a pytorch tensor object or other input is going to be communicated to a generic "pytorch backend" without any special (de)serialization code. Is there some other layer that deserializes the tensor object before passing it to the pytorch model? If so I think that's not clearly detailed.

In this diagram:

![image](https://user-images.githubusercontent.com/11878166/70580345-449e5f80-1b68-11ea-8b9d-baa81a9ad7b0.png)

I would update it so [Serving SDK and Customer Code] is part of TorchServe itself but that is where the (de)serialization code would be. I think most model serving solutions ask for some handler code which contains definitions for init, preprocess, postprocess, and inference functions but in this case we would have some default implementations of above that would be able to convert different input types.

**Comment by MrParosk** [comment]
Hi, is there any updates on this? Is this PR related to multi-model-server (https://github.com/awslabs/multi-model-server)? Is there any proof-of-concept repo / code? (would be keen to help)
